Added support for generic host based IWebHostBuilder

davidfowl opened the pull request:
- This adds an implementation of IWebHostBuilder as a facade over the IHostBuilder.
This removes the 2 container issue by executing the Startup.ConfigureServies and Startup.ConfigureContainer inline as part of building the IHostBuilder.
- The implementation is highly compatible implementation since it exposes the same IWebHostBuilder interface.
Existing extensions mostly work.
- There are some caveats with this approach.
    - Injecting services into Startup is not extremely constrained to the
    services availble on HostBuilderContext. This includes the IHostingEnvironment
    and the IConfiguration.
    - IStartup is broken when using this pattern because it isn't composable.
    - The IStartupConfigureServicesFilter and IStartupConfigureContainer The before
    and after filters added in 2.1 are also broken  because there's a single container (it could maybe be fixed by downcasting and doing something specific on the GenericHostBuilder instance) (@javiercn)
    - Calling into IWebHostBuilder.Build will throw a NotSupportedException since
    this implementation is just a facade over the IHostBuilder.
    - TestServer isn't compatible with this implementation as yet.

Fixes #1544

There are 3 incompatibilities found in running the existing test suite:

- [x] - HostingContextContainsAppConfigurationDuringConfigureLogging => Hosting configuration flows to ConfigureLogging call backs application configuration does not (Won't fix)
- [x] - HostingContextContainsAppConfigurationDuringConfigureServices =>  Hosting configuration flows to ConfigureServices callbacks not application configuration does not (Won't fix)
- [x] - ExternalContainerInstanceCanBeUsedForEverything => Adding an IServiceProviderFactory as a service no longer works. Call IHostBuilder.UseServiceProviderFactory instead. (Won't fix)

This is what we're aiming for with the 3.0 template:

```C#
using System.Threading.Tasks;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Hosting;

namespace SampleApplication
{
    public class Program
    {
        public static async Task Main(string[] args)
        {
            await CreateHostBuilder(args).Build().RunAsync();
        }

        public static IHostBuilder CreateHostBuilder(string[] args) =>
            Host.CreateDefaultBuilder(args)
                .ConfigureWebHostDefaults(builder =>
                {
                    builder.UseStartup<Startup>();
                });
    }
}
```

davidfowl committed:
Added support for generic host based IWebHostBuilder
- This adds an implementation of IWebHostBuilder as a facade over the IHostBuilder.
This removes the 2 container issue by executing the Startup.ConfigureServies and Startup.ConfigureContainer inline as part of building the IHostBuilder.
- The implementation is highly compatible implementation since it exposes the same IWebHostBuilder interface.
Existing extensions mostly work.
- There are some caveats with this approach.
    - Injecting services into Startup is not extremely constrained to the
    services availble on HostBuilderContext. This includes the IHostingEnvironment
    and the IConfiguration.
    - IStartup is broken when using this pattern because it isn't composable.
    - The IStartupConfigureServicesFilter and IStartupConfigureContainer The before
    and after filters added in 2.1 are also broken  because there's a single container (it could maybe be fixed by downcasting and doing something specific on the GenericHostBuilder instance).
    - Calling into IWebHostBuilder.Build will throw a NotSupportedException since
    this implementation is just a facade over the IHostBuilder.

davidfowl commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostedService.cs:
```diff
@@ -0,0 +1,200 @@
+    {
+        private static readonly string DeprecatedServerUrlsKey = "server.urls";
+
+        public GenericWebHostService(IOptions<GenericWebHostServiceOptions> options,
```
Something interesting here, we should maybe delay resolution of these services until StartAsync. If any of these things fail to resolve, you won't get startup errors in the browser.

rynowak replied:
What of these things aren't registered by hosting?

davidfowl replied:
Wait, I just realized, if the server can't be resolved we can't capture startup errors and do anything meaningful anyways 😄 . This is fine.

davidfowl committed:
Added support for UseStartup(assemblyName) and added a single test (so far)

davidfowl committed:
Small tweaks to fix up some comments

davidfowl committed:
Make IHostingStartup work and respect configuration.
- Added a few more tests

Tratcher commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs:
```diff
@@ -29,15 +29,21 @@ public static IWebHostBuilder Configure(this IWebHostBuilder hostBuilder, Action
+            hostBuilder.UseSetting(WebHostDefaults.ApplicationKey, startupAssemblyName);
+
+            // Light up the GenericWebHostBuilder implementation
+            if (hostBuilder is GenericWebHostBuilder genericWebHostBuilder)
```
Rather than sniffing for a specific type can we at least sniff for a new interface like IConfigureStartup with Configure, UseStartup, etc..

davidfowl replied:
Sure but why?

davidfowl replied:
Nah. It’s overkill and can be added at anytime. I would do that if we had something else concrete that would implement. Also I plan to keep this all internal as we’re not trying to expose anything public.

So in the end we’d have a new internal interface implemented and an internal type that would allow us to test these extension methods. I don’t see the value.

Tratcher replied:
👃

Tratcher replied:
Added them where?

An interface per extension seems overkill as they're all related to Startup/Configure.

davidfowl commented:
This might be the way we recommend frameworks plug into the generic host. A single top level method with a builder specific to that framework. I'm thinking that Orleans could potentially use the same approach with the SiloBuilder.

cc @ReubenBond @jdom

davidfowl committed:
Call HostStop after StopAsync

davidfowl committed:
Renamed the IApplicationLifetime implementation

davidfowl commented:
cc @tillig 

Just an FYI that this breaks the pattern where you return an IServiceProvider from Startup.ConfigureServices. The preferred pattern will be an IServiceProviderFactory. That makes things compose since the web application configuration isn't the only thing existing. This will require a migration story when moving apps from 2.x to 3.x. Luckily, we're not breaking any of the WebHostBuilder behavior so it can be smooth 😄 .

davidfowl committed:
Cleanup the hosted service

tillig commented:
Thanks for the heads-up. I'll make sure our corresponding docs and examples get updated!

davidfowl commented:
Pinging other DI container authors (feel free to spread the word if I missed people) cc @dadhi @seesharper @ipjohnson @khellang @ENikS

davidfowl committed:
Make default application throw when not configured.

davidfowl committed:
Throw good error if no application delegate was specified

davidfowl committed:
Unify the tests for the GenericWebHostBuilder and WebHostBuilder
- There's 3 failures but 2 are a known incompatiblity and the last one is a bug

davidfowl committed:
Fix ordering issues with IHostingStartup
- Put callbacks in place first to run IHostingStartup modifications using a wrapper IWebHostBuilder (now we have 3 of them...).

davidfowl committed:
Added light up interfaces for IWebHostBuilders
- Added ISupportsStartup and ISupportsDefaultServiceProvider that circumvents the default extension method logic and delegates to the type itself.

oferns commented:
@dotnetjunkie might be interested in this conversation

seesharper commented:
@davidfowl Thanks for letting us know up front about changes like this. 

Correct me if I am wrong, but this is what I can gather from this

*  `IServiceProvider  ConfigureServices` will not be supported anymore.
*  We can now add the `IServiceProviderFactory` to an `IHostBuilder` rather that the `IWebHostBuilder`  https://github.com/seesharper/LightInject.Microsoft.AspNetCore.Hosting/blob/master/src/LightInject.Microsoft.AspNetCore.Hosting/LightInjectWebHostBuilderExtensions.cs

davidfowl committed:
Added tests for incompatible behavior in GenericWebHostBuilder

rynowak commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHost.cs:
```diff
@@ -0,0 +1,32 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace Microsoft.AspNetCore.Hosting.Internal
```
Eww `.Internal` - this is just noise

rynowak replied:
Also folder and namespace haven't even met each other.

davidfowl replied:
Yea, I wanted to group the implementation together without sticking .GenericHost in the namespace name.

rynowak replied:
put them in one file and use `#region`

rynowak commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHost.cs:
```diff
@@ -0,0 +1,32 @@
+﻿using System;
```
copyright

rynowak replied:
and everytwhere

rynowak commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHost.cs:
```diff
@@ -0,0 +1,32 @@
+
+        public void Dispose() => _host.Dispose();
+
+        public void Start() => _host.Start();
```
Is this really right? Is the fact that this class is an implementation of `IWebHost` even significant since users can't call `.Build()`?

rynowak replied:
I suppose this is here so you can resolve it as a service, but if you do - is it right to have these methods do what they do?

psibernetic replied:
Why wouldn't webhost just call base on build?

rynowak commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostApplicationLifetime.cs:
```diff
@@ -0,0 +1,21 @@
+    internal class GenericWebHostApplicationLifetime : IApplicationLifetime
+    {
+        private readonly Microsoft.Extensions.Hosting.IApplicationLifetime _applicationLifetime;
+        public GenericWebHostApplicationLifetime(Microsoft.Extensions.Hosting.IApplicationLifetime applicationLifetime)
```
You from a year ago is trolling you now 😆

rynowak commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostedService.cs:
```diff
@@ -0,0 +1,197 @@
+{
+    internal class GenericWebHostService : IHostedService
+    {
+        private static readonly string DeprecatedServerUrlsKey = "server.urls";
```
If this is deprecated do we need it?

Tratcher replied:
Yes this should be removed.

davidfowl replied:
Removed it

rynowak commented on src/Microsoft.AspNetCore.Hosting/GenericHost/HostingStartupWebHostBuilder.cs:
```diff
@@ -0,0 +1,78 @@
+
+        public IWebHost Build()
+        {
+            throw new NotSupportedException();
```
Message?

rynowak commented on src/Microsoft.AspNetCore.Hosting/GenericHost/HostingStartupWebHostBuilder.cs:
```diff
@@ -0,0 +1,78 @@
+    {
+        private readonly GenericWebHostBuilder _builder;
+        private Action<WebHostBuilderContext, IConfigurationBuilder> _configureConfiguration;
+        private Action<WebHostBuilderContext, IServiceCollection> _configureServices;
```
Is it worth clearing these after use? Or will this create more problems?

davidfowl committed:
Added copyright and added error message text

javiercn commented:
> * The IStartupConfigureServicesFilter and IStartupConfigureContainer The before
>   and after filters added in 2.1 are also broken  because there's a single container (it could maybe be fixed by downcasting and doing something specific on the GenericHostBuilder instance) (@javiercn)

I think I know how we can fix this easily. We can use the new properties dictionary in IHostBuilderContext (should be an abstract class) to tag the filter instances in there and then when ConfigureServices is invoked on the host builder, we simply retrieve the filters from the hostbuildercontext and apply them.

Tratcher commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostBuilder.cs:
```diff
@@ -0,0 +1,382 @@
+                services.TryAddTransient<IServiceProviderFactory<IServiceCollection>, DefaultServiceProviderFactory>();
+
+                // Ensure object pooling is available everywhere.
+                services.TryAddSingleton<ObjectPoolProvider, DefaultObjectPoolProvider>();
```
Now is a great time to drop it before anyone takes a dependency on it.

davidfowl replied:
We just need routing to add the ObjectPoolProvider https://github.com/aspnet/Routing/blob/7b16053f27ac982eb71317fa8072b55c443aa158/src/Microsoft.AspNetCore.Routing/DependencyInjection/RoutingServiceCollectionExtensions.cs#L36.

davidfowl committed:
Remove things

Tratcher commented on src/Microsoft.AspNetCore.Hosting/WebHostBuilderExtensions.cs:
```diff
@@ -51,8 +57,15 @@ public static IWebHostBuilder UseStartup(this IWebHostBuilder hostBuilder, Type
         {
```
What about the overload here:
https://github.com/aspnet/Hosting/blob/2de091e154cc33a5b3d413572826c523d98462a4/src/Microsoft.AspNetCore.Hosting.Abstractions/HostingAbstractionsWebHostBuilderExtensions.cs#L52-L63

davidfowl replied:
What about it? This works fine.

davidfowl replied:
https://github.com/aspnet/Hosting/pull/1580/files#diff-575646ff9c6b4006e3d4eb641c7c8f01R105

davidfowl replied:
https://github.com/aspnet/Hosting/blob/111d6932a07cb732b304e04025348deba4d2ed75/src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostBuilder.cs#L105

Tratcher commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostBuilder.cs:
```diff
@@ -0,0 +1,385 @@
+        }
+
+        // This exists just so that we can use ActivatorUtilities.CreateInstance on the Startup class
+        private class ServiceProvider : IServiceProvider
```
HostServiceProvider?

davidfowl commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostBuilder.cs:
```diff
@@ -0,0 +1,385 @@
+
+namespace Microsoft.AspNetCore.Hosting.Internal
+{
+    internal class GenericWebHostBuilder : IWebHostBuilder, ISupportsStartup, ISupportsUseDefaultServiceProvider
```
Do you have any actionable feedback? Do you see a leak?

Tratcher replied:
It does leak in that it lives beyond the obvious lifespan from the extension method, but since there's only one that shouldn't cause too much trouble. I don't have a specific suggestion here, I was only calling this out as an unusual side-effect of this design approach.

davidfowl commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostBuilder.cs:
```diff
@@ -0,0 +1,385 @@
+        private AggregateException _hostingStartupErrors;
+        private HostingStartupWebHostBuilder _hostingStartupWebHostBuilder;
+
+        public GenericWebHostBuilder(IHostBuilder builder, bool allowBuild = true)
```
Why does that matter? Nobody can create this class as it's internal.

Tratcher replied:
Trying to limit the amount of test code mixed into production code, and to prevent mistakes like defaulting to a test configuration.

Tratcher replied:
https://github.com/aspnet/Hosting/pull/1586

davidfowl committed:
Renamed ServiceProvider -> HostServiceProvider

Tratcher commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostedService.cs:
```diff
@@ -0,0 +1,200 @@
+            try
+            {
+                var builder = ApplicationBuilderFactory.CreateBuilder(Server.Features);
+                Action<IApplicationBuilder> configure = Options.ConfigureApplication ?? _defaultApplication;
```
Nevermind, I misread the signature of _defaultApplication, I thought it was throwing a per-request exception.  It's a little strange you wouldn't just throw the exception here rather than running the filters first. Do we have an examples of terminal filters that would would replace the app, or error handling filters that would catch the exception?

Tratcher replied:
This scenario wasn't supported by the old WebHost. If you didn't specify startup then the filters never ran. https://github.com/aspnet/Hosting/blob/d7b9fd480765bdc01f06441f308fb288e6001049/src/Microsoft.AspNetCore.Hosting/Internal/WebHost.cs#L191-L196

davidfowl replied:
So this behavior is important enough that a test should fail right?

tillig commented:
> Just an FYI that this breaks the pattern where you return an IServiceProvider from Startup.ConfigureServices. The preferred pattern will be an IServiceProviderFactory.

Currently the only way to build the Autofac container _and keep a reference to it_ if you need to do that is to return the `IServiceProvider` from `Startup.ConfigureServices`. We've seen some interesting use cases where people want to use a single container and host multiple application types (like a self-hosted ASP.NET Core app alongside... uh... something else? Not sure what). That's led to requests for things like the ability to create a pseudo-container in a child lifetime scope.

```
+-------------------------------------------------------+
| root container: shared registrations                  |
| and singletons                                        |
|                                                       |
| +------------------------+ +------------------------+ |
| | child scope: root      | | child scope: root      | |
| | for ASP.NET Core self- | | for background service | |
| | hosted app             | | or something (?)       | |
| +------------------------+ +------------------------+ |
+-------------------------------------------------------+
```

Having the ability to return the `IServiceProvider` means the dev could provide a pre-created scope to the `ConfigureServices` method and return a custom-built `IServiceProvider`.

I gather that in order to do something like this in the new system folks will need a custom `IServiceProviderFactory` or figure out a workaround.

Tratcher commented on src/Microsoft.AspNetCore.Hosting/GenericHost/GenericWebHostBuilder.cs:
```diff
@@ -0,0 +1,385 @@
+
+                // We do this super early but still late enough that we can process the configuration
+                // wired up by calls to UseSetting
+                ExecuteHostingStartups();
```
In the WebHostBuilder world you could configure HostingStartup via the host config APIs (UseSetting, UseConfiguration). Those still work for generic web host, but only if you call them directly inside the web host context. As implemented, HostingStartup will not honor host level configuration set via generic hosts's ConfigureHostConfiguration. This is the only feature that ends up using an isolated config.

Mitigation: Before calling ExecuteHostingStartups, call `_config = config.Build()` to at least capture the host config added so far.

poke commented:
Is there any way we can bring back the functionality that allows us to inject an logger into the `Startup`? It’s very common for applications to log certain things during startup, at a time when the DI container has not been built.

I understand that with this implementation, there is only hard-coded support for `IHostingEnvironment` and `IConfiguration` but I think `ILogger<T>` or `ILoggerFactory` would be another of the few types that actually make sense to get injected there.

I’ve already looked into the `HostBuilder` source for this but unfortunately, logging is just set up [through DI](https://github.com/aspnet/Hosting/blob/05d1a6eb0e051a0cbadc64246f77fe9b836d33a6/src/Microsoft.Extensions.Hosting/HostBuilder.cs#L184) so it will be difficult to get a proper logger factory out of it. But I think logging would be a really valuable addition to those explicit types that are supported in the startup.

Tratcher commented:
@poke I don't see how it's possible to bring logging back, it fundamentally depends on DI and config.  It's worth noting that logging was already broken (multiple singletons) and one of the reasons we dropped the second container.

poke commented:
Yeah, I understand that it’s very difficult and not possible in the current design, but I really think that it’s something that is expected to work. So if we don’t end up having a good approach for that, I am sure people will make up a way to achieve this, for example by creating a new (second) logger factory from within their startup. Just like people built the service provider within the `ConfigureServices` method when they needed some random service to configure stuff before it was properly communicated (and supported by utility methods) how to do that properly.

Tratcher commented:
At least logging can still be injected into Configure, it's only ConfigureServices that's missing it.

davidfowl commented:
Logging can't really be done because logger providers are themselves in the DI container.

Tratcher committed:
Seperate test and production code (#1586)

PureKrome commented:
So if we are logging when we `ConfigureServices` which takes an `ILogger`, then we would need to create a new (Second) logger factory (as @poke said) ?

davidfowl committed:
Don't run startup filters if the application isn't configured

davidfowl commented:
> So if we are logging when we ConfigureServices, where a method (in here) takes an ILogger, then we would need to create a new (Second) logger factory (as @poke said) ?

If you're logging today in ConfigureServices then you're in this torn state where you're using another instance of the ILoggerFactory that's configured. What that can lead to is several background threads running for each of the logger factories (for example the console logger and blob storage logger).

Sure. you can create another logger factory that is separate in there but where would you be logging to?

davidfowl commented:
@tillig Yes, you'd need a custom factory to do that, but you can capture the instance by using the Configure method.

Tratcher reviewed (approved):
Update https://github.com/aspnet/Hosting/blob/master/samples/GenericWebHost/ and then I think you're good for now.

davidfowl commented:
> Update https://github.com/aspnet/Hosting/blob/master/samples/GenericWebHost/ and then I think you're good for 

😍

davidfowl committed:
Use new extension method in sample

PureKrome commented:
> > So if we are logging when we ConfigureServices, where a method (in here) takes an ILogger, then we would need to create a new (Second) logger factory (as @poke said) ?
> 
> If you're logging today in ConfigureServices then you're in this torn state where you're using another instance of the ILoggerFactory that's configured. What that can lead to is several background threads running for each of the logger factories (for example the console logger and blob storage logger).
> 

yep. but can we dispose of this factory (and therefore any bg threads) once we leave the scope? Then there could be issues of flushing, etc ?

> Sure. you can create another logger factory that is separate in there but where would you be logging to?

Console / some 3rd party service 

in the example above i have: 

`services.AddLogging(configure => configure.AddSerilog(dispose: true));`

so the logging sinks are read-in/defined -prior- to the `HostBuilder` getting created. So logging has been added/configured for serilog ... and then I was hoping to use this seril-logger as a MEL `ILogger`. Happy that there's no DI in this bit .. but possible to expose/use currently existing one?

poke commented:
> Logging can't really be done because logger providers are themselves in the DI container.

I know but I think there still _has_ to be some way to do it. Even if you don’t necessarily need to log stuff out from the `ConfigureServices` method, people will still want to do it (there’s a relatively popular [SO question](https://stackoverflow.com/q/41287648/216074) on this that proves the desire).

And I really don’t want people to end up doing things like this:
```
public void ConfigureServices(IServiceCollection services)
{
    var provider = services.BuildServiceProvider();
    var logger = provider.GetService<ILogger<Startup>>();

    logger.LogInformation("Adding MVC");
    services.AddMvc();
}
```
But that’s totally going to happen if we don’t have an actual recommendation on what to do instead.

You mentioned on Slack that you are considering a root logger as a singleton, and I think that would be totally fine. Just allow _some_ sane way to get a logger inside of there, so we can direct people to that solution.

davidfowl commented:
That will be the way to do it regardless, it *will* require another container. It should not be the same service provider that you use the build up the application though. We'll discuss it more.

poke commented:
Yeah, but since there already is a separate service provider at web host level by default, it just works. It even uses the same configuration and setup even if it’s technically a separate instance. And that would be my expectation with 3.0 too: That there would be a default and easy way that just works, even if this means that the framework will create a separate logger for me.

davidfowl commented:
"Just works" There's been numerous unsolvable bugs around the fact that we have 2 containers (sometimes 3!). Most of our workarounds have been to not log anything until after the application container is wired up. We want to be out of that world.

davidfowl commented:
cc @ENikS I can't remember if I mentioned you on this.
